### PR TITLE
feat(search-option): add lock status search option

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

## Description

This PR introduces a new search option for displaying the lock status of items in GLPI. The lock status is displayed as an icon, with additional details available in a tooltip on hover. This simplifies the user interface and makes it easier to quickly understand the lock status of an item.

image

## Changes

- Addition of a new `lock_status.html.twig` template for the lock icon and tooltip. This template uses conditional operators to set the color, icon, and tooltip based on the `is_locked` variable.
- Modification of `ObjectLock.php` to add a new search option for lock status and to use the new template. The `getSpecificValueToDisplay` method was added to handle the display of the lock status.

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
To improve accessibility maybe we could add some text using a badge ?

image

**Comment by ccailly** [comment]
> To improve accessibility maybe we could add some text using a badge ?
> 
> image

The style changes have been applied, and I've modified the PR description.

**Review by AdrienClairembault — APPROVED** [review]

File: src/ObjectLock.php [hunk 1]
```diff
@@ -621,11 +623,41 @@ public static function rawSearchOptionsToAdd($itemtype)
...
 
+    public static function getSpecificValueToDisplay($field, $values, array $options = [])
+    {
+        if (!is_array($values)) {
+            $values = [$field => $values];
+        }
+
+        switch ($field) {
+            case 'id':
+                return TemplateRenderer::getInstance()->render('components/search/lock_status.html.twig', [
```
Comment (orthagh) [inline src/ObjectLock.php:651]: For a quick html code, I'm not sure we gain something to have dedicated template, unless you plan to reuse it.

Comment (AdrienClairembault) [reply src/ObjectLock.php:651]: Always better than keeping some raw `echo` statement, no ?

Comment (cedric-anne) [reply src/ObjectLock.php:651]: You can reuse this piece of code from #15688:

https://github.com/glpi-project/glpi/blob/0488b363681edf721ce4897964f4a8e2a9836123/src/Application/View/TemplateRenderer.php#L193-L203

File: src/ObjectLock.php [hunk 2]
```diff
@@ -621,11 +623,53 @@ public static function rawSearchOptionsToAdd($itemtype)
...
+                    {% set color = is_locked ? 'bg-red-lt' : 'bg-green-lt' %}
+                    {% set icon = is_locked ? 'ti-lock' : 'ti-lock-open' %}
+                    {% set text = is_locked ? __('Locked') : __('Free') %}
+                    {% set tooltip = is_locked ? __('Locked by %s at %s')|format(call('getUsername', [users_id]), date) : text %}
+
+                    <span class="badge {{ color }}" data-bs-toggle="tooltip" title="{{ tooltip }}">
+                        <i class="ti {{ icon }}"></i>
+                        {{ text }}
+                    </span>
+                TWIG;
```
Comment (cedric-anne) [inline src/ObjectLock.php:661]:
```suggestion
TWIG;
```

File: src/ObjectLock.php [hunk 3]
```diff
@@ -621,11 +623,53 @@ public static function rawSearchOptionsToAdd($itemtype)
...
+            $values = [$field => $values];
+        }
+
+        switch ($field) {
+            case 'id':
+                $templateContent = <<<TWIG
+                    {% set color = is_locked ? 'bg-red-lt' : 'bg-green-lt' %}
+                    {% set icon = is_locked ? 'ti-lock' : 'ti-lock-open' %}
+                    {% set text = is_locked ? __('Locked') : __('Free') %}
+                    {% set tooltip = is_locked ? __('Locked by %s at %s')|format(call('getUsername', [users_id]), date) : text %}
```
Comment (cedric-anne) [inline src/ObjectLock.php:655]: Call to `getUsername()` can be done on PHP side.
```suggestion
                    {% set tooltip = is_locked ? __('Locked by %s at %s')|format(user_name, date) : text %}
```

File: src/ObjectLock.php [hunk 4]
```diff
@@ -621,11 +623,53 @@ public static function rawSearchOptionsToAdd($itemtype)
...
+
+                    <span class="badge {{ color }}" data-bs-toggle="tooltip" title="{{ tooltip }}">
+                        <i class="ti {{ icon }}"></i>
+                        {{ text }}
+                    </span>
+                TWIG;
+
+                return TemplateRenderer::getInstance()->renderFromStringTemplate($templateContent, [
+                    'is_locked' => $values['id'] > 0,
+                    'users_id'  => $values['users_id'],
```
Comment (cedric-anne) [inline src/ObjectLock.php:665]:
```suggestion
                    'user_name' => getUsername($values['users_id']),
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

File: src/ObjectLock.php [hunk 5]
```diff
@@ -621,11 +623,53 @@ public static function rawSearchOptionsToAdd($itemtype)
...
+
+                    <span class="badge {{ color }}" data-bs-toggle="tooltip" title="{{ tooltip }}">
+                        <i class="ti {{ icon }}"></i>
+                        {{ text }}
+                    </span>
+TWIG;
+
+                return TemplateRenderer::getInstance()->renderFromStringTemplate($templateContent, [
+                    'is_locked' => $values['id'] > 0,
+                    'user_name' => getUsername($values['users_id']),
```
Comment (trasher) [inline src/ObjectLock.php:665]: `getUsername()` function does not exists.

**Review by trasher — CHANGES_REQUESTED** [review]

File: src/ObjectLock.php (hunk 5 above)
Comment (trasher) [reply src/ObjectLock.php:665]: Probably:
```suggestion
                    'user_name' => getUserName($values['users_id']),
```

**Review by trasher — APPROVED** [review]

**Pull request merged**